<h1 style = "text-align: center; font-size: 48px;">MOVIE RATING PREDICTION WITH PYTHON<h1>

<h2>Introduction</h2>

<p>
Movies are without a doubt, an essential part of the modern age. Each Movie has a story to tell, an idea to share, or a case to defend.
This dataset is pulled from IMDb.com of all the Indian movies on the platform. <br>
Our task today will be to clean this data by different method, and to provide some usefull insight into how to the dataset is partitioned, accompagned by studying the relationship between different Features.
All of this will be assembled into what we call, A Data Preprocessing Pipeline. More on that later on.
</p>

<h2>Page Index</h2>
<ol>
    <li>Data collection</li>
    <li>Data visualization</li>
    <li>Data processing</li>
    <li>Model initialisation</li>
    <li>Valuation and optimization of the prediciton module</li>
</ol>

<h2>Data Collection</h2>

<h3>Source</h3>

The data has been collected from the famous website <i>Kaggle</i> which contains many usefull data set and informations.<br>
Link : <a href="https://www.kaggle.com/datasets/adrianmcmahon/imdb-india-movies">IMDb India Movies</a>

<h3>Libraries</h3>

In [1]:
import pandas as pd
import seaborn as sns
import numpy as np
import matplotlib.pyplot as plt
import sklearn as skl

In [2]:
print('Pandas version is : ', pd.__version__)
print('NumPy version is : ', np.__version__)
print('Seaborn version is : ', sns.__version__)
print('matplotlib version is : ', plt.matplotlib.__version__)
print('sklearn version is : ', skl.__version__)


Pandas version is :  2.0.3
NumPy version is :  1.24.3
Seaborn version is :  0.12.2
matplotlib version is :  3.7.2
sklearn version is :  1.3.0


<h3>Importing the dataset</h3>

In [3]:
df = pd.read_csv('IMDb Movies India.csv')

In [4]:
df.head()

Name    Year Duration            Genre  \
0                                         NaN      NaN            Drama   
1  #Gadhvi (He thought he was Gandhi)  (2019)  109 min            Drama   
2                         #Homecoming  (2021)   90 min   Drama, Musical   
3                             #Yaaram  (2019)  110 min  Comedy, Romance   
4                   ...And Once Again  (2010)  105 min            Drama   

   Rating Votes            Director       Actor 1             Actor 2  \
0     NaN   NaN       J.S. Randhawa      Manmauji              Birbal   
1     7.0     8       Gaurav Bakshi  Rasika Dugal      Vivek Ghamande   
2     NaN   NaN  Soumyajit Majumdar  Sayani Gupta   Plabita Borthakur   
3     4.4    35          Ovais Khan       Prateik          Ishita Raj   
4     NaN   NaN        Amol Palekar  Rajat Kapoor  Rituparna Sengupta   

           Actor 3  
0  Rajendra Bhatia  
1    Arvind Jangid  
2       Roy Angana  
3  Siddhant Kapoor  
4      Antara Mali

<h3>Understanding the dataset</h3>

In [5]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 15509 entries, 0 to 15508
Data columns (total 10 columns):
 #   Column    Non-Null Count  Dtype  
---  ------    --------------  -----  
 0   Name      15509 non-null  object 
 1   Year      14981 non-null  object 
 2   Duration  7240 non-null   object 
 3   Genre     13632 non-null  object 
 4   Rating    7919 non-null   float64
 5   Votes     7920 non-null   object 
 6   Director  14984 non-null  object 
 7   Actor 1   13892 non-null  object 
 8   Actor 2   13125 non-null  object 
 9   Actor 3   12365 non-null  object 
dtypes: float64(1), object(9)
memory usage: 1.2+ MB


<h3>Data description</h3>

<div style="width: 90%;margin:auto; background-color: #fff; padding: 20px; border-radius: 8px; box-shadow: 0 0 5px rgba(0, 0, 0, 0.1);">
    <h1 style="text-align: center; color: #333;">Movie Rating Dataset Description</h1>
    <p>In our dataset, we observe the following:</p>
    <ol>
        <li>There are a total of <span style="font-weight: bold; color: #555;">15,509</span> entries/rows.</li>
        <li>Each movie entry includes <span style="font-weight: bold; color: #555;">10</span> different features:</li>
    </ol>
    <ul>
        <li><span style="font-weight: bold; color: #555;">Name</span>: The title of the movie (string value).</li>
        <li><span style="font-weight: bold; color: #555;">Year</span>: The year in which the movie was released (integer).</li>
        <li><span style="font-weight: bold; color: #555;">Duration</span>: The length of the movie in minutes (float).</li>
        <li><span style="font-weight: bold; color: #555;">Genre</span>: The genre of the movie (string).</li>
        <li><span style="font-weight: bold; color: #555;">Rating</span>: A rating on a scale from 0 to 10, indicating viewer satisfaction (float).</li>
        <li><span style="font-weight: bold; color: #555;">Votes</span>: The number of people who have rated the movie.</li>
        <li><span style="font-weight: bold; color: #555;">Director</span>: The director of the movie.</li>
        <li><span style="font-weight: bold; color: #555;">Actor 1</span>: The main actor of the movie.</li>
        <li><span style="font-weight: bold; color: #555;">Actor 2</span>: The second main actor of the movie.</li>
        <li><span style="font-weight: bold; color: #555;">Actor 3</span>: The third main actor of the movie.</li>
    </ul>
</div>


<h3>Data Preprocessing</h3>

<p>
One of the most common way to handle any dataset in order to feed it to a prediction model or any sort of models, is to build what we call a 'Data Pipeline'.<br>
This will guide us into building a much cleaner and more presentable study case, and to ensure our models are as efficient and noise free as possible. <br>
    First, let's analyse in depth our dataset:
</p>

In [6]:
df.isna().sum()

Name           0
Year         528
Duration    8269
Genre       1877
Rating      7590
Votes       7589
Director     525
Actor 1     1617
Actor 2     2384
Actor 3     3144
dtype: int64

Here we notice some intresting stuffs:<br>
    + 528 movies don't have their year of publishement.<br>
    + We are missing the duration for 8269 movies!.<br>
    + 1877 Movies are missing their genre.<br>
        .<br>
        .<br>
        .<br>
    <br>

We also notice that the Feature 'Genre' contains entries which consist of multiple genre, it's better to decompose each indivduals Genre into a Feature and reporesent the abscence of existance of a Genre by binary data: 1 for yes and 0 for no.

## Genres list

In [7]:
df_genre = df.copy()
df_genre.dropna(subset = 'Genre', axis = 0, inplace = True)
genres = []
genres = set(genre for genre_list in df_genre['Genre'].str.split(',') for genre in genre_list)
print(genres)

{' Romance', 'Animation', ' Crime', 'War', ' Sport', 'Mystery', 'Reality-TV', ' Action', 'Comedy', ' Music', 'Fantasy', 'Romance', 'Music', ' Biography', ' News', 'Thriller', 'Horror', 'Biography', 'Action', 'History', ' Reality-TV', 'Drama', ' Musical', ' Horror', ' Adventure', ' War', ' Mystery', 'Sci-Fi', 'Sport', ' Family', 'Adventure', ' Sci-Fi', 'Documentary', ' Western', ' Comedy', ' Short', 'Musical', 'Family', 'Crime', ' Fantasy', ' History', ' Drama', ' Thriller'}


## Pipeline 

In [8]:
def imputate(dataset):
    '''Since we are looking to predict the rating of a movie, it's better to keep only rows with valid 'Rating' entries
    instead of just filling in with some variable.
    For the genre, we'll keep only the main genre of the film, suppossing that the first one in order is the main one.
    '''
    
    dataset.dropna(subset= ['Rating', 'Votes'], axis = 0, inplace = True)

    dataset.dropna(subset = 'Year', axis = 0 , inplace = True)

    df['Genre'] = df['Genre'].str.split(',', expand=True)[0]
    return dataset


def format_dataset(dataset):
    # Remove parentheses from the 'Year' column and convert to integer
    dataset['Year'] = dataset['Year'].str.strip('()').astype(int)
    dataset['Duration'] = dataset['Duration'].str.rstrip(' min').astype(float)
    dataset['Rating'] = dataset['Rating'].astype(float)
    dataset['Votes'] = dataset['Votes'].astype(float)
    dataset['Name'] = dataset['Name'].str.extract('([A-Za-z\s\'\-]+)')
    return dataset



def preprocessing(dataset):
    
    imputate(dataset)
    format_dataset(dataset)
    
    X = dataset.drop('Rating', axis = 1)
    Y = dataset['Rating']
    return X, Y


In [9]:
preprocessing(df)
df

Name  Year  Duration   Genre  Rating   Votes  \
1                  Gadhvi   2019     109.0   Drama     7.0     8.0   
3                   Yaaram  2019     110.0  Comedy     4.4    35.0   
5        Aur Pyaar Ho Gaya  1997     147.0  Comedy     4.7   827.0   
6                   Yahaan  2005     142.0   Drama     7.4  1086.0   
8          A Question Mark  2012      82.0  Horror     5.6   326.0   
...                    ...   ...       ...     ...     ...     ...   
15501      Zulm Ki Hukumat  1992       NaN  Action     5.3   135.0   
15503      Zulm Ki Zanjeer  1989     125.0  Action     5.8    44.0   
15504  Zulm Ko Jala Doonga  1988       NaN  Action     4.6    11.0   
15505                Zulmi  1999     129.0  Action     4.5   655.0   
15508         Zulm-O-Sitam  1998     130.0  Action     6.2    20.0   

                Director           Actor 1                 Actor 2  \
1          Gaurav Bakshi      Rasika Dugal          Vivek Ghamande   
3             Ovais Khan           Prateik              Ishita Raj   
5           Rahul Rawail        Bobby Deol  Aishwarya Rai Bachchan   
6         Shoojit Sircar   Jimmy Sheirgill          Minissha Lamba   
8          Allyson Patel         Yash Dave          Muntazir Ahmad   
...                  ...               ...                     ...   
15501  Bharat Rangachary        Dharmendra     Moushumi Chatterjee   
15503    S.P. Muthuraman       Chiranjeevi              Jayamalini   
15504      Mahendra Shah  Naseeruddin Shah           Sumeet Saigal   
15505         Kuku Kohli      Akshay Kumar          Twinkle Khanna   
15508       K.C. Bokadia        Dharmendra              Jaya Prada   

               Actor 3  
1        Arvind Jangid  
3      Siddhant Kapoor  
5        Shammi Kapoor  
6       Yashpal Sharma  
8         Kiran Bhatia  
...                ...  
15501          Govinda  
15503      Rajinikanth  
15504    Suparna Anand  
15505      Aruna Irani  
15508      Arjun Sarja  

[7919 rows x 10 columns]

<h3>Data's Format</h3>

In [10]:
# for a in df.columns:
#     column_format = df1[a].apply(lambda x: type(x)).unique()

#     if len(column_format) == 1:
#         print(f"All values in the '{a}' column have the same format: {column_format[0]}")
#     else:
#         print(f"Values in the '{a}' column have different formats: {column_format}")